cmd/relayer/validate.go

```diff
@@ -0,0 +1,156 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/athanorlabs/atomic-swap/coins"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+	rcommon "github.com/athanorlabs/go-relayer/common"
+
+	"github.com/cockroachdb/apd/v3"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+var (
+	uint256Ty, _ = abi.NewType("uint256", "", nil)
+	bytes32Ty, _ = abi.NewType("bytes32", "", nil)
+	addressTy, _ = abi.NewType("address", "", nil)
+	arguments    = abi.Arguments{
+		{
+			Name: "owner",
+			Type: addressTy,
+		},
+		{
+			Name: "claimer",
+			Type: addressTy,
+		},
+		{
+			Name: "pubKeyClaim",
+			Type: bytes32Ty,
+		},
+		{
+			Name: "pubKeyRefund",
+			Type: bytes32Ty,
+		},
+		{
+			Name: "timeout0",
+			Type: uint256Ty,
+		},
+		{
+			Name: "timeout1",
+			Type: uint256Ty,
+		},
+		{
+			Name: "asset",
+			Type: addressTy,
+		},
+		{
+			Name: "value",
+			Type: uint256Ty,
+		},
+		{
+			Name: "nonce",
+			Type: uint256Ty,
+		},
+		{
+			Name: "_s",
+			Type: bytes32Ty,
+		},
+		{
+			Name: "fee",
+			Type: uint256Ty,
+		},
+	}
+)
+
+type validator struct {
+	ctx               context.Context
+	ec                *ethclient.Client
+	relayerCommission *apd.Decimal
+	forwarderAddress  ethcommon.Address
+}
+
+func (v *validator) validateTransactionFunc(req *rcommon.SubmitTransactionRequest) error {
+	// validate that:
+	// 1. the `to` address is a swap contract;
+	// 2. the function being called is `claimRelayer`;
+	// 3. the fee passed to `claimRelayer` is equal to or greater
+	// than our desired commission percentage.
```

Comment:
Is there any easy way we can validate that the commission on the swap amount will exceed the commission fees paid by the relayer? (Maybe in a follow-up PR, if not here?)

One idea would be to handle it by setting a minimum fee to use the relayer.  Another option would be to have a minimum ETH amount for a swap. (These options are not mutually exclusive.)  On the monero side, we definitely need to institute a minimum swap amount to avoid ugly support questions. If the swap amount is too small, the monero gets trapped in the A+B wallet, as the wallet doesn't have enough funds to sweep any portion of the swap amount out of it.

Reply:
yeah, we'd have to add something in the relayer code that would allow for a gas estimate to happen before actual execution, then we could find the max gas fee from there. however since the call should always take the same amount of gas (for ETH, ERC20s would need to be estimated probably) we can determine the price just by multiplying with the gas price.

I think having a fixed relayer commission could also be a solution. there's not really any "attack" possible on relayers thankfully, since instantiating the swap costs significantly more than calling `claim()` does.

overall, there probably should be a minimum swap amount, the amount should at least be higher than tx fees like you said